### PR TITLE
Exportable preprocessing and on device training

### PR DESCRIPTION
**Pull request by tfiedlerdev**
In order to make model inference more efficient and easier in our android app, we added a NormalizationLayer to the RainBowModel class which CAN be utilized by child classes. 
We also automatically compute and store the feature-wise variance and mean for each dataset using DataConfig. 
These measures are then implicitly exported into the model as parameters of the normalization layer. 
 We also export function signatures for training the model, inferring, ... on device. These methods use tensorflow functions to apply forward fill on the raw windows presented to the model in order to remove NaNs. This reduces code required in the app.

### CODE REVIEW

File: src/utils/DataConfig.py [hunk 1]
```diff
@@ -27,11 +32,43 @@ class DataConfig:
...
+    DATA_CONFIG_METADATA_FILE = "dataConfigMetadata.json"
 
-    # interface (subclass responsibility to define) ------------------------------------------------------------
+    def __init__(self, dataset_path: str):
+        self.dataset_path = dataset_path
 
     def load_dataset(self) -> "list[Recording]":
+        recordings = self._load_dataset()
+
+        print(
```
Comment (lucasliebe) [inline src/utils/DataConfig.py:46]: i am a bit confused by the prints here, you state that calculating might take a while before checking if they are available, then you only print that they where found if you need to calculate them and no message is written if they have been set before

File: src/utils/DataConfig.py [hunk 2]
```diff
@@ -99,16 +194,19 @@ def __init__(self, dataset_path: str):
             5: "sandwich time",
         }
 
-    def load_dataset(self) -> "list[Recording]":
+    def _load_dataset(self) -> "list[Recording]":
         return load_opportunity_dataset(self.dataset_path)
 
+    def _getDataConfigIdentifier(self):
```
Comment (lucasliebe) [inline src/utils/DataConfig.py:200]: same method as in parent class, why do we need to declare it again?

File: src/utils/DataConfig.py [hunk 3]
```diff
@@ -240,19 +339,28 @@ def load_dataset(self, **args) -> "list[Recording]":
         },
     ]
 
+    def _getDataConfigIdentifier(self):
```
Comment (lucasliebe) [inline src/utils/DataConfig.py:342]: same method as in parent class, why do we need to declare it again?

File: src/utils/DataConfig.py [hunk 4]
```diff
@@ -240,19 +339,28 @@ def load_dataset(self, **args) -> "list[Recording]":
...
+
+    category_labels = {'rollstuhl transfer': 0, 'essen reichen': 1, 'umkleiden': 2, 'bad vorbereiten': 3, 'bett machen': 4, 'gesamtwaschen im bett': 5, 'aufräumen': 6, 'geschirr einsammeln': 7, 'essen austragen': 8, 'getränke ausschenken': 9, 'küchenvorbereitung': 10,
+                       'waschen am waschbecken': 11, 'rollstuhl schieben': 12, 'mundpflege': 13, 'haare kämmen': 14, 'essen auf teller geben': 15, 'dokumentation': 16, 'aufwischen (staub)': 17, 'haare waschen': 18, 'medikamente stellen': 19, 'accessoires anlegen': 20, 'föhnen': 21}
 
-    category_labels = {'rollstuhl transfer': 0, 'essen reichen': 1, 'umkleiden': 2, 'bad vorbereiten': 3, 'bett machen': 4, 'gesamtwaschen im bett': 5, 'aufräumen': 6, 'geschirr einsammeln': 7, 'essen austragen': 8, 'getränke ausschenken': 9, 'küchenvorbereitung': 10, 'waschen am waschbecken': 11, 'rollstuhl schieben': 12, 'mundpflege': 13, 'haare kämmen': 14, 'essen auf teller geben': 15, 'dokumentation': 16, 'aufwischen (staub)': 17, 'haare waschen': 18, 'medikamente stellen': 19, 'accessoires anlegen': 20, 'föhnen': 21}
+    def _getDataConfigIdentifier(self):
```
Comment (lucasliebe) [inline src/utils/DataConfig.py:365]: same method as in parent class, why do we need to declare it again?

**Review by lucasliebe — CHANGES_REQUESTED** [review]
Please check my comments and feel free to discuss with me about some of them

File: src/models/RainbowModel.py [hunk 5]
```diff
@@ -59,37 +46,115 @@ def __init__(self, **kwargs):
...
+        return np.array(sensor_arrays), np.array(activity_vectors)
+
+    # The 'train' function takes an input window and a label
+    @tf.function(input_signature=[
+        tf.TensorSpec(shape=[None, window_size, n_features], dtype=tf.float32),
+        # Binary Classification
+        tf.TensorSpec(shape=[None, n_outputs], dtype=tf.float32),
+    ])
+    def train(self, input_windows, labels):
+        replaceNaN_ffill_tf(input_windows)
```
[the remark this answers is not in the document]
Comment (tfiedlerdev) [reply src/models/RainbowModel.py:127]: Preprocessing has not been done before here. 
This is one of the exportable function signatures that one will be calling on device. As we don't want to do preprocessing in Kotlin anymore, we're doing it with the raw input_windows that we receive from kotlin

File: src/models/RainbowModel.py [hunk 6]
```diff
@@ -98,56 +163,42 @@ def fit(self, X_train: np.ndarray, y_train: np.ndarray) -> None:
...
-
-
+        self.history = history
 
     # Predict ------------------------------------------------------------------------
+    @tf.function(input_signature=[
+        tf.TensorSpec(shape=[None, window_size, n_features], dtype=tf.float32),
+    ])
+    def infer(self, input_window):
+        replaceNaN_ffill_tf(input_window)
```
[the remark this answers is not in the document]
Comment (tfiedlerdev) [reply src/models/RainbowModel.py:194]: Same answer as for `train`

File: src/utils/DataConfig.py (hunk 2 above)
Comment (tfiedlerdev) [reply src/utils/DataConfig.py:200]: Historisch gewachsen

**Review by tfiedlerdev — COMMENTED** [review]
Quickly skimmed through everything. We might need to remove some unused imports (and the archives!!!!) in the end but I think we can merge this for now

**Review by lucasliebe — APPROVED** [review]
This branch is supposed to be the new main state of the repository so it can be merged into master. Many developers also were included in development. My previous comments also where adopted so I am satisfied with this.

**Pull request merged**